### PR TITLE
Capability discovery

### LINKED ISSUE #1519: Capability Discovery
## Abstract

A mechanism for capability discovery is proposed. Capability discovery eases compatibility and discovery concerns related to operating in a heterogeneous network where nodes support a mix of features. The technical mechanism is as follows:
* A [bitstring](https://en.wikipedia.org/wiki/Bit_array) enumerates the orchestrator's supported and broadcaster's required features
* An additional message to specify non-binary constraints such as a numeric range of acceptable parameters. We do not currently require any constraints, but the basic mechanism is sketched here for completeness.
* The broadcaster filters for orchestrators satisfying its requirements during the discovery phase, matching on the bitstring and comparing against the constraints.

## Motivation

When a new feature is added that requires support on both sides of the B-O network boundary [1] (recent examples: MP4 support, non-integer frame rates, durations), then consideration needs to be given towards compatibility. A robust compatibility mechanism is important to minimize the impact of changes, to maximize the reach of the network, and to provide a framework for achieving such  while minimizing engineering effort. The criteria for a compatibility system are as follows:

1. How to identify nodes that have upgraded to support a new feature
2. How to identify nodes that might not support the new feature, whether due to being non-upgraded or having the feature disabled
3. How to interoperate with non-upgraded nodes when the new feature is not needed

Capability discovery is proposed as a mechanism for this.

[1] The O-T network boundary also has similar issues around compatibility, since transcoders cannot be expected to be upgraded in tandem with orchestrators. The same capability matching mechanism should be re-used during transcoder selection by orchestrators.

### Proposed Solution

Two parts: feature bit-string and capability constraints.

The capabilities of the orchestrator are advertised as a bitstring. Each bit in the string corresponds to a certain capability (or feature). Capabilities using this bitstring mechanism are binary: either they are enabled, or not.

On discovery, the orchestrator sends down its capability bitstring within the `OrchestratorInfo` response. The broadcaster constructs its own (reduced) bitstring corresponding to the transcoding requirements for that particular job. Capability checking of a particular orchestrator corresponds to checking that the AND between the two bitstrings is equivalent to the broadcaster's own capability requirements:

```
isOrchestratorCapable := (broadcaster.bitstring AND orchestrator.bitstring) == broadcaster.bitstring
```

The orchestrator populates its bitstring at startup. Some capabilities might be explicitly toggled via config, others might be auto-detected (eg, codec support if using particular GPUs), and some others might simply be always-on as compatibility markers. The broadcaster generates its bitstring based on job requirements, perhaps during initialization of the orchestrator discovery procedure.

Some features are not binary, so can not be represented as a bitstring. Limits or ranges might be common. For example, perhaps the orchestrator opts to constrain the video resolution, duration or bitrate it's willing to process. Currently, there is no requirement to handle such constraints, but it is certain to occur eventually. For such cases, a separate structure can be used, with each constraint added as a field, and a comparison defined for each particular constraint. See [Capability Constraints](#Implementation_Considerations_Capability_Constraints) for additional details.

For additional details on why the distinction between feature bitstrings and capability constraints is necessary, see [Alternative to Bitstring: Constraint-Only Matching](#Alternative_to_Bitstring_Constraint-Only_Matching).

### Compatibility Criteria

Feature bit-strings address the criteria for compatibility:

1. How to select nodes that have upgraded to support the new feature

Orchestrators that have not upgraded to support a certain feature will have the corresponding bit-indices marked as zero. Non-upgraded nodes will be excluded as a result of this.

2. How to identify nodes that might *not* support the new feature, eg disabling a feature

Orchestrators that have certain features disabled will have the corresponding bit-indices marked as zero.

3. How to interoperate with older nodes when the new feature is not needed

When the job only uses older features, then newer features are not toggled in the broadacster's reduced bitstring of capability requirements. Hence, the bitstring for non-upgraded orchestrators will continue to match the broadcaster's requirements.

##  Implementation Considerations : Bitstring Construction, Matching and Feature Mapping

Each new feature or capability is assigned a permanent bit-index. This bit-index assignment must be shared by all nodes that wish to interoperate. For example, in golang, a fixed set of enumerations could be used:

```golang
// assume these are the features that are enabled, with their bit-indexes
const (
  Capability_MPEGTS = 0
  Capability_MP4 = 1
  // any others formats can be listed here
  Capability_FractionalFramerates = 5
  // any other features can be listed here
  Capability_H264Profile_Baseline = 10
  Capability_H264Profile_Main = 1
  Capability_H264Profile_High = 12
  Capability_H264Profile_ConstrainedHigh = 13
  // and so forth
  Capability_HEVC_Decoding = 20
)
```

For simplicity of representation and processing, it may be suitable to represent the bitstring as a list of uint64 values. Here is how to construct and match on the bitstring using a list-of-uint64 representation.

```golang
// Assume all these features are enabled
features := [Capability_MPEGTS, Capability_MP4, Capability_FractionalFramerates, Capability_H264Profile_Baseline, Capability_HEVC_Decoding]
//  bit string construction
bitstring = make([]uint64, ceil(len(features)/64))
for i := range features {
  bit_index := features[i]
  int_index := floor(bit_index / 64)
  int_bit_index := bit_index % 64
  bitstring[int_index] ||=  (1 << int_bit_index)
}

// broadcaster / orchestrator bit string matching
// assume we have a "reduced bitstring" with job requirements
bcast_bitstring = ... // example of how to map this to job below
if len(bcast_bitstring) > len(orch_bitstring) return false
for i := range bcast_bitstring {
  if bcast_bitstring[i] != orch_bitstring[i] return false 
}
return true
```

Here is an example of how a broadacster's requirements bitstring could be constructed based on the job parameters. This may seem tedious, but the construction is straightforward. One day we could build better abstractions for the construction, if they present themselves. For now, something like the below will likely suffice:

```golang
// broadcaster mapping of feature requirements to bitstring indices
has_mpegts := false
has_mp4 := false
has_h264_baseline := false
// ... and so on for the rest of the features.
for i := range job.outputs {
  has_mpegts ||= job.outputs[i].Format == Format_MPEGTS
  has_mp4 ||= job.outputs[i].Format == Format_MP4
  has_h264_baseline ||= job.outputs[i].H264Profile == Profile_H264Baseline
}
if (has_mpegts) features = append(features, Capability_MPEGTS)
// and so forth
// now invoke the bitsting construction routine described earlier, given `feature`
```

## Implementation Considerations: Capability Constraints

Some features are not binary, so can not be as easily represented in a simple bitstring. Limits or ranges might be common. For example, perhaps the orchestrator opts to constrain the video resolution, duration or bitrate it's willing to process. Currently, there is no requirement to handle such constraints, but something is likely to occur eventually. For completeness, here is a sketch of how such constraint-matching could work.

```proto
message CapabilityConstraints {
  MaxWidth: 1920
  MaxHeight: 1080
  MaxDuration: 600000
}
```

Additional constraints will have to be explicitly checked for each field. It might be enough to also represent the job's constraints within the same type of structure, and compare the two,  eg:

```golang
orchConstraints := orchestratorInfo.CapabilityConstraints
bcastConstraints := ComputeBroadcasterConstraints(segTranscodingMetadata)
useOrchestrator := orchConstraints.IsAcceptable(bcastConstraints)

func ComputeBroadcasterConstraints(md *SegTranscodingMetadata) (*CapabilityConstraints) {
  c := &AdditionalConstraints{}
  
  // Populate source-based fields as needed
  c.Duration = md.Duration
  // ... other fields
  
  // Populate output-based fields as needed
  for _, profile := md.Profiles {
      c.MaxWidth = max(profile.Width, c.MaxWidth)
      // For the sake of example, assume we have a "minimum" constraint. Populate zero value first.
      c.MinWidth = min(profile.Width, c.MinWidth)
      // ... other fields
  }
}

func (orchConstraints *CapabilityConstraints) IsAcceptable(bcastConstraints *CapabilityConstraints) bool {
  if orchConstraints.MaxWidth < bcastConstraints.MaxWidth return false
  // For the sake of an example, assume there is a minimum-type constraint
  if orchConstraints.MinResolution > bcastConstraints.MinResolution return false
  // ... check other fields ...
  return true
}
```

We may want to also signal support for each additional constraint within the bitstring itself. This would primarily be useful for orchestrator matching on broadcasters, in order to detect whether the broadcaster was expecting compatibility with a certain field the orchestrator is unaware of. Unclear whether such signaling would be "always-on", or enabled as needed.

For the initial implemementation, stub capability constraints structs and functions can be defined. Subsequent work on additional features can fill these out as needed.

## Implementation Considerations : Discovery

The signature for the [`GetOrchestrators`](https://github.com/livepeer/go-livepeer/blob/c6b06679fe4adaa204c6d0a9628e184b2fadc0b6/common/types.go#L18) function will need to be updated to take a set of requirements against which to filter orchestrator capabilities.

Currently, capability discovery will be non-interactive: the broadcaster will not transmit its own requirements during the GetOrchestrators call.

The existing predicate check  [predicate check](https://github.com/livepeer/go-livepeer/blob/c90d4e176ec216564c760272a3d5a1b659cb6af9/discovery/db_discovery.go#L123) may be folded into this new mechanism.

## Implementation Considerations: Compatibility with Non-Capability Enabled Orchestrators

Nodes will not be upgraded simultaneously to support capability discovery, so there still needs to be some consideration put towards compatibility with those older nodes.

When the broadcaster's bitstring is generated for a given job, it can be checked whether it is exclusively comprised of "legacy" features. If the job fits within the legacy feature-set and capability information for the orchestrator is missing, then the orchestrator will still be used, provided it passes the other discovery-stage filters.

The existing mechanisms for backwards compatibility will continue to work as needed, such as attempting a MP4 transcode with a mpegts-only orchestrator.

### Attacks

An orchestrator could set an arbitrarily long bitstring to all 1's in order to obtain as much work as possible.

We will not attempt to handle this right now, and take the orchestrator's word for what they do support. Capability discovery is not a substitute for proper verification. There may be cryptographic mechanisms to guard against this type of attack, but that is outside the scope of this initial spec.

On a related note, the lack of message constraints within Protocol Buffers is a bit of a concern. For example, we cannot reject a list more than 100 elements long during deserialization, or excessively large byte buffers. This may lead to undesirable memory or bandwidth usage. Not a new problem with capability discovery, but these types of attacks are worth noting.

The orchestrator is also incentivized to be honest about its supported constraints to ensure the best possible quality-of-service for its users. Being careless about constraints will only lead to poor results, and less work for the orchestrator over the long term. Additionally, the non-interactive aspect of discovery allows for some social pressure around obvious outliers.

### Pricing Menu

Capability discovery has often been discussed alongside a "price menu". While there is some overlap with the ideas behind capability signaling, this proposal does not attempt to specify a granular pricing mechanism.

### Orchestrator Rejection of Broadcasters

This approach addresses compatibility concerns from the broadcaster side - eg, how to select matching orchestrators. Does not address compatibility from the orchestrator side - eg, around using (or rejecting) broadcasters based on their capabilities. This may become necessary eventually, but is not currently in scope.

In general, an orchestrator should still reject work for features it does not support to the extent possible [1]. However, this rejection might come at a later point, such as LPMS erroring out. If applicable, the broadcaster can also take steps to trigger early failures, such as setting a deprecated field to an invalid value. However, with capabilty discovery, the onus is on the broadcaster to select appropriate orchestrators. 

[1] If the broadcaster also includes its capabilities during segment submission, the orchestrator can perform the same capability check that the broadcaster does during discovery.

### Orchestrator Mandatory Capabilities

At some point, the orchestrator may need certain "mandatory capabilities" present on the broadcaster. The absence of a mandatory capability would indicate the broadcaster isn't sufficiently up-to-date for the orchestrator. An example of a mandatory capability might be a change in PM handling that needs to be mirrored on both B / O.

An orchestrator signaling a mandatory capability is essentially a hard break with older B versions, which would be good to avoid as much as possible. Although we don't need mandatory capabilities right now, it might be good to incorporate sooner rather than later, in order to ensure forward-compatibility: older broadcasters can error out during discovery if it receives an unsupported mandatory capability from a newer orchestrator, rather than deferring the failure to segment submission time.

For robustness, the orchestrator should also check the broadcaster's own capabilities if there is anything mandatory, unless there is a way to fail out early at segment submission time.

The check for mandatory capabilities could resemble this:

```
(broadcaster.bitstring AND orchestrator.bitstring) == (broadcaster.bitstring OR orchestrator.mandatories)
```

### DB Discovery

Would be good to add this in at some point as a shortcut during the discovery process. Maybe not in the initial implementation.

### Standalone T

This capability discovery mechanism should be ported to standalone T, where the non-interactivity works well for orchestrators to select a T for the job. However, standalone T might not be in the initial implementation. The transcoder's capabilities can be advertised within its [`RegisterRequest`](https://github.com/livepeer/go-livepeer/blob/dace5d5982bafe3bbfdc67b122192f49a23f46e2/net/lp_rpc.proto#L201-L209).

### Alternative to Bitstring: Constraint-Only Matching

Rather than have a "two-step" matching process (one matching the bitstring, and another running through the constraints) we could simply only run through the constraints, and have each binary capability as a named field in the constraint message. This is a bit problematic for a few reasons:

* Iterating through dozens of otherwise-unused fields can be significantly slower in time and space. The proposed approach takes `ceil(numFeatures / 64)` checks.
* Encoding each field over the wire becomes much more expensive; field numbers 1-15 take a minimum of one byte to encode, two bytes for fields 16-2047. This will impose a significant cost in transmitting and processing each message. On the other hand, bitstrings allow us to encode each binary feature in exactly one bit. https://developers.google.com/protocol-buffers/docs/proto#assigning-field-numbers
* It is harder to act against fields that the node is unaware of (eg, orchestrator rejection of unknown constraints specified by the B). One way to work around this is to encode the presence of each new constraint within the bitstring itself.

### Alternative to Bitstring + Constraints: Version Ratchet

One simple way to upgrade is via a version ratchet, where the orchestrator advertises a network version number (O.N), and the broadcaster only works with orchestrators that are greater than or equal to its own network version (B.N ; `B.N >= O.N` ). https://github.com/livepeer/go-livepeer/pull/1433 .

However, the coarseness of version ratchets is a problem. Criteria # 3 is not met : as soon as a node upgrades its network version, it is cut off from all older nodes (hence, 'ratchet'), shrinking the effective size of the network. Likewise, criteria #2 is not met either, since version ratchets do not allow for signaling whether a certain feature may be available.

## Alternative to Bitstring: Bloom Filter

Alternative: Rather than a bitstring, a mechanism such as Bloom filters can be used. This might only be necessary if the bitstring becomes extremely large. Much of the constraints would have to remain, such as the indices that a certain feature hashes into. Feature matching also becomes linear on the number of features in use, since each feature needs to be looked up. Non-binary features would still have to be handled separately.

## Alternative : Interactive Capability Discovery Protocol

The broadcaster can transmit its capability requirements, and the orchestrator can acknowledge whether it is able to satisfy the request. This may be useful one day for orchestrator routing, eg to direct work towards specific nodes that have hardware qualified for a certain job. However, for now, we'll stay with a non-interactive protocol for the following reasons:
 
- Having a non-interactive advertisement of capabilities is useful for overall network visibility.

- Codifying this non-interactive behavior should also facilitate reciprocal checks later on without inadvertently increasing B / O coupling during discovery

- Non-interactivity makes this more portable to other contexts such as standalone T and orchestrator selection thereof, where there isn't an OrchestratorInfo message transmitted prior to a job starting.
 
- May be difficult for the orchestrator to reject on non-binary constraints that it is unaware of. This type of "detection by omission" is easier for the broadcaster, since the broadcaster knows what it needs to look for. This could perhaps be mitigated by inspecting the message for any leftover, un-serialized data outside the current schema, but that may not be a completely reliable approach. Another workaround is to encode the presence of each additional constraint within the bitstring itself. The longer capability discovery is non-interactive, the longer we can defer addressing these issues.

### PR DESCRIPTION
**Pull request by j0sh**
**What does this pull request do? Explain your changes. (required)**

Implements the first stages of capability discovery - https://github.com/livepeer/go-livepeer/issues/1519

This does not implement (and will not implement):

* Additional constraints. Only stubs out the structures.
* Capability storage in broadcaster DB
* Capability discovery for standalone T

**Specific updates (required)**

- net: Regenerate proto file.  c2aed29
- core: Add support for capability enumeration.  2a3f05a
- discovery: Add capability discovery. cba60a5
- core: Add encoding profiles to capabilities. bef2014 Added to demonstrate how to add a new capability. Note that the list of profiles is explicit, if one ever needs to be disabled or superseded.
- core: Add GOP to capabilities. cb4d6ae Added to demonstrate how to add a new capability.
- core: Add mandatory capability support. efdc188 See https://github.com/livepeer/go-livepeer/issues/1519 for a description of orchestrator mandatory capabilities. Currently, nothing is mandatory, but some things might be soon (eg, transcoding protocol authentication - https://github.com/livepeer/go-livepeer/issues/1569 )
-server: Add orchestrator checking of segment capabilities. … e79ec85 . Acts as a validation check prior to segment processing, and needed for capability matching with standalone T (not currently supported).

**How did you test each of these updates (required)**

* Unit testing
* Regression testing via https://gist.github.com/j0sh/9d884a19bb92e504092e71fd19d812e5
* Manual testing, including  builds with certain capabilities intentionally disabled to induce failures. (There is currently no way to selectively disable capabilities outside of compile time.)
  * Capability Discovery B,  Old O
  * Old B, Capability Discovery O
  * Capability Discovery B, Capability Discovery O, All Features Enabled
  * Capability Discovery B, Capability Discovery O, mpegts disabled, MP4 transcoding request
  * Capability Discovery B, Capability Discovery O, mpegts disabled, mpegts transcoding request
  * [ ] TODO Test a build with object storage disabled, and external OS in use

**Does this pull request close any open issues?**

Fixes https://github.com/livepeer/go-livepeer/issues/1519

**Checklist:**

- [ ] README and other documentation updated
- [ ] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

File: core/capabilities.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+const (
+	Capability_Invalid Capability = iota - 2
+	Capability_Unused
+	Capability_H264
+	Capability_MPEGTS
+	Capability_MP4
+	Capability_FractionalFramerates
+	Capability_StorageDirect
+	Capability_StorageS3
+	Capability_StorageGCS
```
Comment (j0sh) [inline core/capabilities.go:30]: Figured it'd be good to be as specific as possible about all the features that are being used, including the storage driver, in case we ever need to disable support for one of these someday.

**Review by jailuthra — APPROVED** [review]
LGTM!

File: core/capabilities.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	for k, _ := range caps {
+		capList = append(capList, k)
+	}
+
+	return &Capabilities{bitstring: NewCapabilityString(capList)}, nil
+}
+
+func (bcast *Capabilities) CompatibleWith(orch *net.Capabilities) bool {
+	if bcast == nil {
+		// Weird golang behavior: interface value can evaluate to non-nil
```
Comment (kyriediculous) [inline core/capabilities.go:104]: An interface in go will only equal nil if both the type and value are nil , since an interface consist of both a type and a value.

**Comment by kyriediculous** [comment]
Just lurking here 👀  

Uttermost interested in "Capability discovery for standalone T"

Should we consider gitignoring the generated grpc go stubs because they're part of the makefile anyhow (I notice a conflict) ?

**Comment by j0sh** [comment]
> Should we consider gitignoring the generated grpc go stubs

@kyriediculous Do you mean the `net/lp_rpc.pb.go` file ? That would require the protobuf-go compiler be installed - it's a bit easier sometimes without that. Also, different versions of the compiler can cause unwanted changes or even breakage as @jailuthra encountered recently [1]. Checking in the generated code allows us to punt on the issue and ensure everything is synchronized for everyone.

[1] We should definitely still fix these types of issues eventually, maybe at the same time we fix go 1.14 😅  https://github.com/livepeer/go-livepeer/pull/1414

**Comment by j0sh** [comment]
Rebased against master and added a few more things.

Profiles and GOP support (separately) to demonstrate how to add new capabilities.

Mandatory capabilities because this allows for "forward compatibility" - even though we don't have any such mandatory capabilities right now, broadcasters that know to check the field will be able to reject orchestrators that add mandatory capabilities in the future, even if the broadcasters themselves are not aware of the capability. See https://github.com/livepeer/go-livepeer/issues/1519 for more details.

Orchestrator checking of job capabilities - because this is going to be needed at some point, and it's a good sanity check on the orchestrator side.

Still needs a little more manual testing, especially around object storage, but things are ready for review. This PR is as complete as it's going to get, feature-wise.

File: core/capabilities.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+	// set expected storage
+	storageCap, err := storageToCapability(params.OS)
+	if err != nil {
+		return nil, err
+	}
+	caps[storageCap] = true
+
+	// generate bitstring
+	capList := []Capability{}
+	for k, _ := range caps {
```
Comment (yondonfu) [inline core/capabilities.go]: Should be able to shorten this to `for k := range caps`

File: core/capabilities.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+	if bcast == nil {
+		// Weird golang behavior: interface value can evaluate to non-nil
+		// even if the underlying concrete type is nil.
+		// cf. common.CapabilityComparator
+		return false
+	}
+
+	// For now, check this:
+	// ( orch.mandatories  AND orch.bitstring  ) == orch.mandatories &&
```
Comment (yondonfu) [inline core/capabilities.go]: Is this comment accurate? Based on the checks below I would think that this comment should say:

```
( orch.mandatories AND bcast.bitstring ) == orch.mandatories &&
( bcast.bitstring AND orch.bitstring ) == bcast.bitstring
```

File: server/rpc.go [hunk 5]
```diff
@@ -340,12 +342,20 @@ func coreSegMetadata(segData *net.SegData) (*core.SegTranscodingMetadata, error)
 		dur = 2 * time.Second // assume 2sec default duration
 	}
 
+	caps := core.CapabilitiesFromNetCapabilities(segData.Capabilities)
+	if caps == nil {
+		// For older broadcasters. Note if there are any orchestrator
+		// mandatory capabilities, segment verification will fail.
```
Comment (yondonfu) [inline server/rpc.go]: Should "segment verification will fail" say something like "seg creds verification will fail" since I believe it is referring to the compatibility check in `verifySegCreds()`?

**Review by yondonfu — COMMENTED** [review]
Looks great just a few minor comments.

File: core/capabilities.go (hunk 3 above)
Comment (j0sh) [reply core/capabilities.go]: Fixed in 176c77e0d5ec5b7cd30e1cb13746ee16728d9513

File: core/capabilities.go (hunk 4 above)
Comment (j0sh) [reply core/capabilities.go]: Yep, typos. Fixed in 192981f3893798f656f0ea973d4b45c48a2e63a6

File: server/rpc.go (hunk 5 above)
Comment (j0sh) [reply server/rpc.go]: Fixed in 2eca7be58e2990dabf716372c7734bee1e7d992f

**Review by yondonfu — APPROVED** [review]
LGTM :shipit:
